packages/arb-token-bridge-ui/src/token-bridge-sdk/teleport.ts

```diff
@@ -55,3 +60,87 @@ export const fetchTeleportStatusFromTxId = async ({
     l3Provider: destinationChainProvider
   })
 }
+
+async function isContract(
+  address: string,
+  provider: providers.Provider
+): Promise<boolean> {
+  const code = await provider.getCode(address)
+  return code !== '0x'
+}
+
+async function tryGetInboxFromRouter(
+  address: string,
+  provider: providers.Provider
+): Promise<string | undefined> {
+  if (!(await isContract(address, provider))) {
+    throw new Error('Not a contract')
+  }
+
+  const maybeRouter = L1GatewayRouter__factory.connect(address, provider)
+
+  try {
+    return await maybeRouter.inbox()
+  } catch (e: any) {
+    if (e.code === 'CALL_EXCEPTION') {
+      return undefined
+    }
+    throw e
+  }
+}
+
+async function getChainIdFromInbox(
+  address: string,
+  provider: providers.Provider
+): Promise<BigNumber> {
+  const inbox = IInbox__factory.connect(address, provider)
+  const bridge = IBridge__factory.connect(await inbox.bridge(), provider)
+  const rollup = IRollupCore__factory.connect(await bridge.rollup(), provider)
+  return rollup.chainId()
+}
+
+async function getChainIdFromInboxOrRouter(
+  address: string,
+  provider: providers.Provider
+): Promise<BigNumber> {
+  const maybeInbox = await tryGetInboxFromRouter(address, provider)
+  if (maybeInbox) {
+    return getChainIdFromInbox(maybeInbox, provider)
+  }
+
+  return getChainIdFromInbox(address, provider)
+}
+
+const cache: { [id: string]: number } = {}
+export const getL3ChainIdFromTeleportEvents = async (
```

Comment:
We need to detect the destination -`l3ChainId` for ERC20 teleports and we do that using the `l2l3RouterOrInbox` or `l1l2Router` information we get from subgraph.